### PR TITLE
Avoid unnecessary Locale copying in ToStringPrecisionLocale

### PR DESCRIPTION
**Pull request by biodranik**
Clean up and minor speed up of the locale-formatting code.

### CODE REVIEW

File: platform/localization.cpp [hunk 1]
```diff
@@ -16,29 +16,29 @@ enum class MeasurementType
   Altitude
 };
 
-const LocalizedUnits & GetLocalizedUnits(measurement_utils::Units units, MeasurementType measurementType)
+LocalizedUnits const & GetLocalizedUnits(Units units, MeasurementType measurementType)
 {
-  static LocalizedUnits UnitsLenghImperial = {GetLocalizedString("ft"), GetLocalizedString("mi")};
-  static LocalizedUnits UnitsLenghMetric = {GetLocalizedString("m"), GetLocalizedString("km")};
+  static LocalizedUnits const UnitsLenghImperial = {GetLocalizedString("ft"), GetLocalizedString("mi")};
```
Comment (RedAuburn) [inline platform/localization.cpp:21]: `UnitsLengh` -> 'UnitsLength'
may as well while touching this stuff :)

**Comment by biodranik** [comment]
@gpesquero please review and test the changes and tell what you think.

File: platform/measurement_utils.cpp [hunk 2]
```diff
@@ -50,15 +45,15 @@ string ToStringPrecisionLocale(Locale loc, double d, int pr)
   {
     // Value with no decimals. Check if it's equal or bigger than 10000 to
     // insert the grouping (thousands) separator characters.
-    if (out.size() > 4 && !loc.m_groupingSeparator.empty())
```
Comment (gpesquero) [inline platform/measurement_utils.cpp:53]: This check `if (out.size() > 4 &&` is needed to add the grouping separators to values equal or bigger than 10000. If we remove it, then also values from 1000 to 9999 will have the grouping separator.

Do we really want this? In the past it was decided to add the grouping separators with values >= 10000.

File: platform/measurement_utils.cpp [hunk 3]
```diff
@@ -18,24 +17,20 @@
...
+  return ToStringPrecisionLocale(locale, d, pr);
 }
 
-string ToStringPrecisionLocale(Locale loc, double d, int pr)
+std::string ToStringPrecisionLocale(platform::Locale const & loc, double d, int pr)
 {
-  stringstream ss;
-  ss << setprecision(pr) << fixed << d;
-  string out = ss.str();
+  // TODO(AB): Check if snprintf is faster.
```
Comment (gpesquero) [inline platform/measurement_utils.cpp:30]: I've made some preliminary tests and `std::snprintf` seems to be much faster than `std::ostringstream`.
I'd keep the current implementation for this PR and launch an additional PR for the `std::snprintf` solution.

File: platform/measurement_utils.cpp (hunk 2 above)
Comment (biodranik) [reply platform/measurement_utils.cpp]: Thanks for reminding about that, I've reverted the change. Do you remember that reasoning? :)

Comment (gpesquero) [reply platform/measurement_utils.cpp]: Here's the reasoning that we followed last year for the grouping separator: https://github.com/organicmaps/organicmaps/pull/4029#discussion_r1297801741

**Comment by gpesquero** [comment]
> @gpesquero please review and test the changes and tell what you think.

@biodranik: This PR looks good to me. I've tested it in Android and iOS and everything seems to work Ok.

In any case, these changes may be modified in the near future due to the following pending tasks:

* Implementation of `std::snprintf` in ToStringPrecisionLocale().
* Additional benchmarking of cpp+jni vs Java of locale formattting ([PR 7955](https://github.com/organicmaps/organicmaps/pull/7955)).
* Enable the change of locale format at runtime ([PR 7779](https://github.com/organicmaps/organicmaps/pull/7779)).
* Adding extra distance & speed formatting testing ([PR 6470](https://github.com/organicmaps/organicmaps/pull/6470)).

**Comment by biodranik** [comment]
@gpesquero thanks! Let's do it step by step. Here are several obvious issues with unnecessary copying, so the code should be better than before. Then move to your PRs. Which order do you think is better?

@vng PTAL

**Comment by gpesquero** [comment]
> @gpesquero thanks! Let's do it step by step. Here are several obvious issues with unnecessary copying, so the code should be better than before. Then move to your PRs. Which order do you think is better?

As you say, let's do it step-by-step, and *maybe* in this order:
1. Merge current PR.
2. Implement, test & benchmark `std::snprintf` vs `std::ostringstream`.
3. Additional benchmarking of cpp+jni vs Java of locale formattting (PR 7955).
4. Enable the change of locale format at runtime (PR 7779).
5. Adding extra distance & speed formatting testing (PR 6470).

**Review by vng — APPROVED** [review]

**Comment by biodranik** [comment]
@gpesquero yes, let's move on like this. It's a known issue that C++ `iostreams` are slow with a lot of overhead inside. For simple string formatting, `sprintf` is a better solution if you know the max length in advance.

**Pull request merged**